### ISSUE TITLE
Cannot format a ProDOS discs

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#125]
**BerliOS Bug #17707**

**Date:** 2010-November-10 19:20
**Submitted By:** georgejh
**Category:** none
**Bug Group:** none
**Assigned To:** warthurton
**Priority:** 4
# 

Hi,

Error "DISK II DRIVE TOO SLOW" appears when attempting to format a ProDOS disk.

I am using ProDOS 1.1.1 Utility disk on S6 D1 and a copy of Master.dsk (DOS) on S6 D2. Error appears when attempting to format new volume on S6 D2.

Best Regards

**Comment by tomcw** [comment]
Reproduced with AppleWin 1.26.3.0:
- Boot prodos_utility.dsk (which is ProDOS 1.1.1)
- 'F' (ProDOS Filer)
- 'V' (Volume commands)
- 'F' (Format a volume)
- Create a blank (empty, zero byte) blank.dsk image into drive-2
- Slot-6, Drive-2, any volume name and get a "DISK II DRIVE TOO SLOW" error

**Comment by tomcw** [comment]
Here's my (long!) analysis and proposed solution for this issue...

ProDOS's FORMATTER writes tracks as:

16 sectors of:
. Gap1(sector0) or Gap3(sector1-15)
. Address Field
. Gap2
. Data Field

Gap1 and Gap2 are fixed sizes of 129 and 5 lots of 40-cycle 0xFF sync bytes.
Gap3 starts as a large size of 30 x 40-cycle 0xFF sync bytes.
If the track just written can't be read back successfully (eg. due to Gap3 being too big and the track overwriting itself), then Gap3 is reduced by 1 and the the track is re-written.
NB. There can be a small bit of track overwriting as up to ~0x7C bytes of Gap1 could be overwritten leaving ~5 0xFF sync bytes before sector0's Address Field.

If the track just written can be read back successfully, but Gap3 >= 0x1B+1(*1) then the FORMATTER will return the error: "DISK II DRIVE TOO SLOW"

```
; Succeeded writing track, now check gap3_count tolerance
79B5-   A4 D4       LDY   $D4		; w_Gap3Count
79B7-   CC 1F 7D    CPY   $7D1F		; kMinGap3 (=0x0E)
79BA-   B0 05       BCS   $79C1
79BC-   A9 04       LDA   #$04		; 
79BE-   4C F9 79    JMP   $79F9		; error code = 4 (Drive too FAST)
;
79C1-   CC 20 7D    CPY   $7D20		; kMaxGap3 (=0x1B)
79C4-   90 05       BCC   $79CB
79C6-   A9 03       LDA   #$03		; w_Gap3Count >= kMaxGap3
79C8-   4C F9 79    JMP   $79F9		; error code = 3 (Drive too SLOW)
;
79CB-	; OK ...
```

(*1) The function (to write Gap1/Gap3 & the Address Field) takes a parameter of Gap3, but actually writes Gap3+1 sync bytes.

So the original problem with this GH issue was that the track is written with too high a Gap3 count. This was true for both .dsk/.nib images and both authentic/enhanced disk access speed.

Initial results were in getting format working for .dsk/authentic mode by improving the precision of the spin emulation. Essentially, calculate the cycle delta between drive latch r/w accesses. If cycle-delta > 40 cycles, then for every 32 cycles, move the track index on by 1. Not totally accurate due to the 40-cycle sync bytes, but good enough.

This improved spin-precision only allowed format to work for authentic/.dsk images though, and not for:
- .nib images
- .dsk images in enhanced disk access speed

The reason for .nib not working is to do with the track lengths:
- DSK: 6384 (0x18F0)
- NIB: 6656 (0x1A00)

Because .nib has a longer track buffer, then it can fit a whole track even when using the too large Gap3 sizes.

NB. .nib images have a large track to (eg) accommodate 18-sector track formats from  Br0derbund, etc. AFAIKT, .nib's orginate from a tool called EDD (and a spin-off called SST), used for imaging a copy-protected disk, sending the data over a modem and reconstructing (by writing to a slow drive).

The reason for enhanced disk access speed not working is that there's no spin emulation in this  mode: when ProDOS writes a sector, there's spin (of 1 disk nibble) between sectors. This doesn't exist in this enhanced mode, so the track can be written completely using the too large Gap3 sizes. For this to work in this mode, the track would have to be ~0x18B0 in size.

So we need a solution that still supports existing .nib images (with tracks of 0x1A00 bytes), but also allows tracks when written to appear to wrap at 0x18B0.

My current working solution is to monitor the written bytes, detect when all 16 sectors have been written (ie. a full track), then to clamp the track length to 0x18B0. I do this by keeping the last 0x18B0 bytes, and zeroing the remaining bytes.

NB. Zero bytes can't exist on the disk, so the ProDOS read routine will loop until it gets a byte with bit7 set - so it skips the zeroed out part of the track buffer and starts reading real data again at the start of the shortened track. (In DOS3.3/ProDOS there's no cycle counting for this, so the code is completely unaware of this extra delay.)

This method works for all 4 combinations of .dsk/.nib x authentic/enhanced.

So what about DOS 3.3's INIT command?
This is similar to ProDOS in that it writes a track with ever decreasing Gap3 sizes until it can successfully read the track its just written.
The differences are:
- Start with Gap3=46 and decrease by 2 each time
- There isn't a max allowed Gap3 size, so if it works with Gap3=46 then that's fine

I will submit a PR so that others can check my code.

**Comment by tomcw** [comment]
### DOS 3.3 INIT bug

qkumba confirmed (on csa2) that this is a bug in DOS 3.3's INIT:

1) Boot with a DOS 3.3 master (eg. Jan 1, 1983). 
2) Insert a non-DOS 3.3 floppy in drive-2 (eg. a ProDOS formatted floppy) 
3) CATALOG,D2 

DISK VOLUME 254 


I/O ERROR 

4) Eject the drive-2 floppy, and insert a previously formatted DOS 3.3 disk 
5) INIT HELLO 

I/O ERROR 

What I'm consistently seeing is this I/O ERROR! 
But it appears that I can flush out some bad DOS 3.3 state(?) by just doing: 
6) CATALOG,D1 
then.. 
7) INIT HELLO,D2 
...now works! 

qkumba said:
>  I consider this to be a bug in DOS. 
What's happening is that the VTOC sector is cached at $B3BC and expected to have sane values.  DOS uses the VTOC to know where to expect the catalog sectors.  Reading a non-DOS disk causes the cache to be filled with garbage values, which triggers the I/O error after the init is complete, but before the catalog sector is written. 
Refreshing the cache by reading a valid DOS disk is what causes it to work again. 

---

And also I asked:
> I also get similar INIT-failing behaviour (again using an emulator/AppleWin) if I use a completely blank (all zeros) 140K .dsk image. 

[qkumba] Yes, that's expected. 
> Now when I run the same test as above: 
> 
> 3) CATALOG,D2 

... 
> No "I/O ERROR" this time for the CATALOG, but VTOC (track $11) is invalid, so again does what you say ("cache filled with garbage values") explain the I/O ERROR? 

[qkumba] DOS thinks that the catalog is empty because the pointer in the VTOC is also zero. 
That's why there's no I/O error, but it means that it fails to fetch any catalog sectors. 
Then it stores the volume number into what it thinks is the VTOC, but later uses that value ($FE) as a sector number, which triggers the I/O error.

**Comment by tomcw** [comment]
Test PR #507 using Copy II+ 9.1:

Format both DOS3.3 and ProDOS using: authentic or enhanced, .dsk or .nib:
Gap1 = 0x102, Gap2 = 0x08, Gap3 = 0x0B (11), TrackSize = 0x18E7

All formatted OK... then use 'Send to CiderPress' to validate image file.

**Comment by tomcw** [comment]
### Explain how the track size of 0x18B0 is derived

When there's no spin (enhanced=1), then 0x18F0 is still too long:
EG. Gap1 = 0x81, Gap3 = 0x1D (29), TrackSize = 0x1963
- 0x1963 - 0x81(Gap1) = 0x18E2, which is < 0x18F0
  - NB. Take off the gap1, since these are just sync, and the track can validate without any Gap1!

Need Gap3 to be 0x1A...
Gap1 = 0x81, Gap3 = 0x1C (28), TrackSize = 0x1954
Gap1 = 0x81, Gap3 = 0x1B (27), TrackSize = 0x1945
Gap1 = 0x81, Gap3 = 0x1A (26), TrackSize = 0x1936
- 0x1936 - 0x81(gap1) = 0x18B5
So need a track size of 0x18B0 (rounding down)

**Comment by tomcw** [comment]
Closing as TODO list is complete.

**Comment by tomcw** [comment]
@tilleul - FYI, this is fixed with AppleWin 1.27.0.0.
